core/vmenv.go

```diff
@@ -0,0 +1,221 @@
+	for {
+		var (
+			context vm.Context
+			err error
+			out []byte
+			address common.Address
+		)
+		context, err = callOrCreate(self)
+		if err == nil {
+			out, address, err = self.execute(context)
+			context.Finish()
+		}
+		if err != machineBrokenError {
+			return out, address, err
+		} else {
+			// ?? mybe will better to switch to the classic vm and try again before ??
+			panic(err)
+		}
+	}
+}
+
+func (self *vmEnv) Call(sender common.Address, to common.Address, data []byte, gas, price, value *big.Int) ([]byte, error) {
+	out, _, err := self.do(func(e *vmEnv)(vm.Context,error){
+		return e.machine.Call(e.number.Uint64(),sender,to,data,gas,price,value)
+	})
+	return out, err
+}
+
+func (self *vmEnv) Create(caller state.AccountObject, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error) {
+	return self.do(func(e *vmEnv)(vm.Context,error){
+		return e.machine.Create(e.number.Uint64(),caller.Address(),data,gas,price,value)
+	})
+}
+
+var machineBrokenError = errors.New("machine broken")
+
+func (self *vmEnv) execute(ctx vm.Context) ([]byte,common.Address,error) {
+	for {
+		req := ctx.Fire()
+		if req != nil {
+			switch req.ID {
+			case vm.RequireAccount:
+				acc := self.db.GetAccount(req.Address)
+				addr := acc.Address()
+				nonce := acc.Nonce()
+				balance := acc.Balance()
+				ctx.CommitAccount(addr,nonce,balance)
+			case vm.RequireCode:
+				addr := req.Address
+				code := self.db.GetCode(addr)
+				hash := self.db.GetCodeHash(addr)
+				ctx.CommitCode(addr,hash,code)
+			case vm.RequireHash:
+				number := req.Number
+				hash := self.hashfn(number)
+				ctx.CommitBlockhash(number,hash)
+			case vm.RequireRules:
+				number := new(big.Int).SetUint64(req.Number)
+				fork := self.fork
+				ctx.CommitRules(req.Number,self.rules.GasTable(number),fork)
+			default:
+				if ctx.Status() == vm.RequireErr {
+					// ?? unsupported VM implementaion ??
+					// should we panic or use known VM instead?
+					panic("unsupported VM RequireError occured")
+				} else {
+					// ?? incorrect VM implementation ??
+					// Fire or Step can return nil or vm.Require only!
+					panic("incorrect VM state")
+				}
+			}
+		} else {
+			switch ctx.Status() {
+			case vm.ExitedOk:
+				out, err := ctx.Out()
+				if err != nil {
+					return nil, nil, machineBrokenError
+				}
+				address, err := ctx.Address()
+				if err != nil {
+					return nil, nil, machineBrokenError
+				}
+				accounts, err := ctx.Accounts()
+				if err != nil {
+					return nil, nil, machineBrokenError
+				}
+				// applying state here
+				for _, v := range accounts {
+					var o state.AccountObject
+					address := v.Address()
+					if self.db.Exist(address) {
+						o = self.db.GetAccount(address)
+					} else {
+						o = self.db.CreateAccount(address)
+					}
+					o.SetBalance(v.Balance())
+					o.SetNonce(v.Nonce())
+					code := v.Code()
+					if code != nil {
+						o.SetCode(v.CodeHash(),code)
```

Comment:
I think this logic is sort-of wrong? Code should only be set when a new account is created?